### PR TITLE
Mixed join condition with function `in`

### PR DESCRIPTION
**Pull request by lgbo-ustc**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix (experimental) unequal join with subqueries for sets which are in the mixed join conditions.

Fixed issue #64513

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with Aarch64
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 1565aec14281de2069fc18dbe6594cea8e813e66 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64775/1565aec14281de2069fc18dbe6594cea8e813e66/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by vdimir — APPROVED** [review]
👍

**Comment by vdimir** [comment]
CI @ 1565aec14281de2069fc18dbe6594cea8e813e66

> Integration tests (aarch64) [5/6] — fail: 1, passed: 581 [Details](https://s3.amazonaws.com/clickhouse-test-reports/64775/1565aec14281de2069fc18dbe6594cea8e813e66/integration_tests__aarch64__[5_6].html)

`test_checking_s3_blobs_paranoid` https://github.com/ClickHouse/ClickHouse/issues/64451

> Stateless tests (debug, s3 storage) [2/6] — fail: 1, passed: 1061, skipped: 25 [Details](https://s3.amazonaws.com/clickhouse-test-reports/64775/1565aec14281de2069fc18dbe6594cea8e813e66/stateless_tests__debug__s3_storage__[2_6].html)

`03006_join_on_inequal_expression_fast` you may use `ORDER BY ALL`, since `t1.key, t2.key` is not unique.

> Stateless tests (tsan) [2/5] — fail: 1, passed: 1268, skipped: 14 [Details](https://s3.amazonaws.com/clickhouse-test-reports/64775/1565aec14281de2069fc18dbe6594cea8e813e66/stateless_tests__tsan__[2_5].html)

`01676_clickhouse_client_autocomplete` https://github.com/ClickHouse/ClickHouse/issues/50626

File: tests/queries/0_stateless/03006_join_on_inequal_expression_fast.sql.j2 [hunk 1]
```diff
@@ -18,6 +18,7 @@ SELECT t1.*, t2.* from t1 {{ join_type }} JOIN t2 ON t1.key = t2.key and (t1.b +
 SELECT t1.*, t2.* from t1 {{ join_type }} JOIN t2 ON t1.key = t2.key and (t1.a < t2.a) ORDER BY (t1.key, t1.attr, t2.key, t2.attr);
 SELECT t1.*, t2.* from t1 {{ join_type }} JOIN t2 ON t1.key = t2.key and t1.c ORDER BY (t1.key, t1.attr, t2.key, t2.attr); -- { serverError INVALID_JOIN_ON_EXPRESSION }
 SELECT * FROM (SELECT 1 AS a, 1 AS b, 1 AS c) AS t1 {{ join_type }} JOIN (SELECT 1 AS a, 1 AS b, 1 AS c) AS t2 ON t1.a = t2.a AND (t1.b > 0 OR t2.b > 0);
+SELECT t1.*, t2.* FROM t1 {{ join_type }} JOIN t2 ON t1.key = t2.key AND (t1.a=2 OR (t2.a IN (SELECT a FROM t1 WHERE a = 3))) ORDER BY t1.key, t2.key;
```
Comment (vdimir) [inline tests/queries/0_stateless/03006_join_on_inequal_expression_fast.sql.j2:21]: Also need to adjust `reference` respectively

```suggestion
SELECT t1.*, t2.* FROM t1 {{ join_type }} JOIN t2 ON t1.key = t2.key AND (t1.a=2 OR (t2.a IN (SELECT a FROM t1 WHERE a = 3))) ORDER BY ALL;
```

Comment (lgbo-ustc) [reply tests/queries/0_stateless/03006_join_on_inequal_expression_fast.sql.j2]: done

**Comment by lgbo-ustc** [comment]
Is this a known issue ? https://github.com/ClickHouse/ClickHouse/actions/runs/9376838738/job/25821288013?pr=64775
```
Jun 05 07:01:16 /build/src/Storages/S3Queue/S3QueueTableMetadata.h:41:17: error: function 'DB::S3QueueTableMetadata::checkEquals' has a definition with different parameter names [readability-inconsistent-declaration-parameter-name,-warnings-as-errors]
Jun 05 07:01:16    41 |     static void checkEquals(const S3QueueSettings & lhs, const S3QueueSettings & rhs);
Jun 05 07:01:16       |                 ^
Jun 05 07:01:16 /build/src/Storages/S3Queue/S3QueueTableMetadata.cpp:189:28: note: the definition seen here
Jun 05 07:01:16   189 | void S3QueueTableMetadata::checkEquals(const S3QueueSettings & current, const S3QueueSettings & expected)
Jun 05 07:01:16       |                            ^
Jun 05 07:01:16 /build/src/Storages/S3Queue/S3QueueTableMetadata.h:41:17: note: differing parameters are named here: ('lhs', 'rhs'), in definition: ('current', 'expected')
Jun 05 07:01:16    41 |     static void checkEquals(const S3QueueSettings & lhs, const S3QueueSettings & rhs);
Jun 05 07:01:16       |                 ^                                   ~~~                          ~~~
Jun 05 07:01:16       |                                                     current                      expected
Jun 05 07:01:16 68516 warnings generated.
Jun 05 07:01:16 Suppressed 69128 warnings (68515 in non-user code, 613 NOLINT).

```

**Comment by vdimir** [comment]
Fixed in https://github.com/ClickHouse/ClickHouse/pull/64827, but fix is not yet  merged

**Comment by vdimir** [comment]
Now  https://github.com/ClickHouse/ClickHouse/pull/64827 is merged, I'll merge this PR as well

**Pull request merged**